### PR TITLE
feat: strongly typed item methods

### PR DESCRIPTION
**Pull request by tranhl**
### Summary:

This PR adds support for strongly-typed custom item methods via a second generic type parameter to `dynamoose.model`.

### Code sample:
```ts
// 1. Create a custom item method type that extends from `ItemMethods`.
// This is ensures that only functions are present in the type.
interface UserItemMethods extends ItemMethods {
	resetPassword: () => Promise<void>;
}

// 2. Create a type for `Item` as usual.
interface UserItem extends Item {
	id: string;
	name: string;
	age: number;
}

// 3. Create a schema.
const userSchema = new dynamoose.Schema({
	"id": String,
	"name": {
		"type": String,
		"index": {
			"type": IndexType.global
		}
	},
	"age": Number
});

// 4. Pass item & custom method types to `dynamoose.model`
const UserModel = dynamoose.model<UserItem, UserItemMethods>(
	"User",
	userSchema
);

const user = await UserModel.get('id');
// 5. Access item methods with type safety.
user.resetPassword() // () => Promise<void>
```

### GitHub linked issue:

Implements part of #1647 

### Type (select 1):
- [ ] Bug fix
- [x] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement
- [ ] Test added to report bug
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [ ] Yes
- [x] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [ ] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/main/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [ ] I have filled out all fields above

### CODE REVIEW

**Comment by tranhl** [comment]
@fishcharlie Hoping to get your input on my design/implementation of strongly typed item methods. I've created a draft PR for now just in case the design needs rework.

I've done the best I can to make sure that anywhere an `Item` is returned, the `ItemMethods` are now also being returned as well (e.g. after executing a `Scan` or `Query` operation). Please do point out if I've missed anything.

**Comment by tranhl** [comment]
I dedicate any and all copyright interest in this software to the
public domain. I make this dedication for the benefit of the public at
large and to the detriment of my heirs and successors. I intend this
dedication to be an overt act of relinquishment in perpetuity of all
present and future rights to this software under copyright law.

File: packages/dynamoose/lib/Model/index.ts [hunk 1]
```diff
@@ -439,21 +439,21 @@ export class Model<T extends ItemCarrier = AnyItem> extends InternalPropertiesCl
 	// latestTableDetails: DynamoDB.DescribeTableOutput;
 	// pendingTaskPromise: () => Promise<void>;
 	Item: typeof ItemCarrier;
```
Comment (tranhl) [inline packages/dynamoose/lib/Model/index.ts:441]: @fishcharlie Is being able to access the underlying `Item` class from a model part of public interface? If it is not, then there aren't any issues. However if it is, then we'll have to figure out a way to better type this property.

For example, this is possible:

```js
const UserModel = dynamoose.model('User', userSchema)
const user = new UserModel.Item({ id: 'foobar' })
```

This works fine in JS, however typing this correctly in TS has proven challenging. I made a naive attempt at fixing the types: 

```suggestion
	Item: T & typeof ItemCarrier;
```

However, TypeScript understandably isn't happy with this, as the internal `Item` class created within `Model` is not assignable to the generic we pass in:

```
Type 'typeof Item' is not assignable to type 'T & typeof Item'.
  Type 'typeof Item' is not assignable to type 'T'.
    'T' could be instantiated with an arbitrary type which could be unrelated to 'typeof Item'.
```

Haven't been able to find a good solution to this yet.

**Comment by tranhl** [comment]
recheck

**Comment by tranhl** [comment]
recheck

File: packages/dynamoose/lib/General.ts [hunk 2]
```diff
@@ -14,7 +14,9 @@ export type KeyObject = {[attribute: string]: string | number};
 // An item representing a DynamoDB key
 export type InputKey = string | number | KeyObject;
 
-interface ModelItemConstructor<T extends Item> {
+type Expand<T> = {[KeyType in keyof T]: T[KeyType]} & {};
```
Comment (fishcharlie) [inline packages/dynamoose/lib/General.ts:17]: Why the `& {}`?

Also, can you give an example of what this would look like? For some reason this type seems very complex at first glance. It's not obviously clear what it's doing and what it's purpose is. Maybe add a comment above with an example or something? Or a better description?

Comment (fishcharlie) [reply packages/dynamoose/lib/General.ts:17]: Also should this be in a separate file?

File: packages/dynamoose/lib/Item.ts [hunk 3]
```diff
@@ -30,6 +30,14 @@ export interface ItemSettings {
 	type?: "fromDynamo" | "toDynamo";
 }
 
+export interface ItemMethods {
+	[key: string]: Function;
+}
+
+export interface AnyItemMethods {
+	[key: string]: any;
+}
```
Comment (fishcharlie) [inline packages/dynamoose/lib/Item.ts:39]: It's not immediately clear what this is doing. Seems more general than just `AnyItemMethods`. Why do we have to use `any` vs `Function`? If it's item methods, shouldn't that be a function?

File: packages/dynamoose/lib/Model/index.ts (hunk 1 above)
Comment (fishcharlie) [reply packages/dynamoose/lib/Model/index.ts:441]: @tranhl 

> Is being able to access the underlying Item class from a model part of public interface?

It might work. But it's not supported. If someone tries to do that in their code I would point them to this: https://dynamoosejs.com/other/FAQ#can-i-use-an-undocumented-property-class-method-or-function-in-dynamoose.

File: packages/dynamoose/test/types/Item.ts [hunk 4]
```diff
@@ -11,3 +11,5 @@ const shouldPassSaveWithReturnItemCallback = user.save({"return": "item"}, () =>
 
 // @ts-expect-error
 const shouldFailWithInvalidReturnType = user.save({"return": "invalid-return-type"});
+
+const shouldPassCustomMethodAccess = user.resetPassword();
```
Comment (fishcharlie) [inline packages/dynamoose/test/types/Item.ts:15]: Maybe we should also have a test with a method that doesn't exist? We can use `// @ts-expect-error` for that one.

File: packages/dynamoose/test/types/Model.ts [hunk 5]
```diff
@@ -89,7 +95,7 @@ const userSchema = new dynamoose.Schema({
 	"age": Number
 });
 
-export const UserTypedModel = dynamoose.model<User>(
+export const UserTypedModel = dynamoose.model<User, UserMethods>(
```
Comment (fishcharlie) [inline packages/dynamoose/test/types/Model.ts:98]: What happens here if you don't pass in `UserMethods`? Does it still work? I kinda think we should be breaking this out into two different tests. The existing tests are just fine. We should have new tests for these UserMethods that don't modify the old tests.

File: packages/dynamoose/test/types/Model.ts [hunk 6]
```diff
@@ -110,3 +116,6 @@ UserTypedModel.update({"id": "foo"}, {
 UserTypedModel.update({"id": "foo"}, {
 	"$REMOVE":{"age":null}
 });
+
+// @ts-expect-error
+const shouldFailItemCustomMethodAccess = UserTypedModel.resetPassword();
```
Comment (fishcharlie) [inline packages/dynamoose/test/types/Model.ts:121]: So are model methods outside the scope of this PR? This PR only deals with item methods it looks like?

**Review by fishcharlie — CHANGES_REQUESTED** [review]
Overall this looks really good.

Beyond the code level comments I made the only other high level suggestion is that we need documentation for this. All the documentation either exists in the code itself using JSDoc or in [`docs/docs_src`](https://github.com/dynamoose/dynamoose/tree/main/docs/docs_src).

File: packages/dynamoose/test/types/Model.ts (hunk 6 above)
Comment (tranhl) [reply packages/dynamoose/test/types/Model.ts:121]: That's correct, I've decided to address them separately to keep scope focused. Happy to include model methods as well if that suits you?

File: packages/dynamoose/lib/General.ts (hunk 2 above)
Comment (tranhl) [reply packages/dynamoose/lib/General.ts:17]: Good points! This type definitely could use a JSDoc explaining its function & intended use. Also agree that it should be moved to a separate file. Best location for it from what I can see would be `lib/Types.ts`. How does that sound?

File: packages/dynamoose/lib/Item.ts (hunk 3 above)
Comment (tranhl) [reply packages/dynamoose/lib/Item.ts:39]: Fair point. I encountered an issue with TS early on omitting the `ItemMethods` generic when calling `dynamoose.model`, however I believe that requiring custom methods to extend `ItemMethods` addresses this. Let me revisit this and see if it's possible to completely remove it.

File: packages/dynamoose/lib/Model/index.ts (hunk 1 above)
Comment (tranhl) [reply packages/dynamoose/lib/Model/index.ts:441]: Okay perfect 👍 .

It's also mentioned in the TS docs that [documentation should be treated as the source of truth](https://dynamoosejs.com/getting_started/TypeScript#should-i-trust-the-documentation-or-typescript-typings-more), so we can set this aside until `Model.Item` becomes public interface.

File: packages/dynamoose/test/types/Item.ts (hunk 4 above)
Comment (tranhl) [reply packages/dynamoose/test/types/Item.ts:15]: Definitely. Test coverage across edge cases was minimal as I was unsure if you'd be happy with the approach of the PR. Will update/refactor tests to be more comprehensive, and update the documentation as well.

File: packages/dynamoose/test/types/Model.ts (hunk 5 above)
Comment (tranhl) [reply packages/dynamoose/test/types/Model.ts:98]: As part of the interface, it should work. As per [my previous comment](https://github.com/dynamoose/dynamoose/pull/1659#discussion_r1477551646), I will make sure this is covered in tests.

File: packages/dynamoose/lib/General.ts (hunk 2 above)
Comment (fishcharlie) [reply packages/dynamoose/lib/General.ts:17]: `lib/Types.ts` works for me 🎉

File: packages/dynamoose/test/types/Model.ts (hunk 6 above)
Comment (fishcharlie) [reply packages/dynamoose/test/types/Model.ts:121]: @tranhl Ideally I'd like to review both PRs before merging either of them. I don't want to be left in a state where one piece is done and people ask why the other piece wasn't implemented.

You can decide how to best structure it. I'd personally prefer 2 separate PRs so that it's easier to review and the context is easier to understand. Sadly I don't think GitHub has the concept of having one PR depend on another.

**Comment by jimmyn** [comment]
It would be great to have this feature live!

**Comment by tranhl** [comment]
FYI my team has chosen to adopt [DynamoDB Toolbox V1](https://www.dynamodbtoolbox.com/) as solves all of the friction of dealing with the DynamoDB Client. As such I'm going to close this PR and recommend that anyone looking for type-safe DynamoDB use Toolbox instead.
